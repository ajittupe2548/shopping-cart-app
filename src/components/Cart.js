import React, { memo, useCallback } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  closeCart,
  removeFromCart,
  increaseQuantity,
  decreaseQuantity,
  clearCart,
} from "../store/cartSlice";
import styles from "./Cart.module.scss";

const Cart = memo(() => {
  const dispatch = useDispatch();
  const { items, totalAmount, isOpen } = useSelector(state => state.cart);

  const handleCloseCart = useCallback(() => {
    dispatch(closeCart());
  }, [dispatch]);

  const handleRemoveItem = useCallback(
    id => {
      dispatch(removeFromCart(id));
    },
    [dispatch]
  );

  const handleIncreaseQuantity = useCallback(
    id => {
      dispatch(increaseQuantity(id));
    },
    [dispatch]
  );

  const handleDecreaseQuantity = useCallback(
    id => {
      dispatch(decreaseQuantity(id));
    },
    [dispatch]
  );

  const handleCheckout = useCallback(() => {
    alert(`Thank you for your purchase! Total: $${totalAmount.toFixed(2)}`);
    dispatch(clearCart());
    dispatch(closeCart());
  }, [dispatch, totalAmount]);

  return (
    <>
      <div
        className={`${styles.cartOverlay} ${isOpen ? styles.open : ""}`}
        onClick={handleCloseCart}
      ></div>

      <div className={`${styles.cartSidebar} ${isOpen ? styles.open : ""}`}>
        <div className={styles.cartHeader}>
          <h2 className={styles.cartTitle}>Shopping Cart</h2>
          <button className={styles.closeCart} onClick={handleCloseCart}>
            âœ•
          </button>
        </div>

        {items.length === 0 ? (
          <div className={styles.emptyCart}>
            <p>Your cart is empty</p>
          </div>
        ) : (
          <>
            <div className={styles.cartItems}>
              {items.map(item => (
                <div key={item.id} className={styles.cartItem}>
                  <img
                    src={item.image}
                    alt={item.title}
                    className={styles.cartItemImage}
                    loading="lazy"
                    decoding="async"
                  />
                  <div className={styles.cartItemDetails}>
                    <h4 className={styles.cartItemTitle}>{item.title}</h4>
                    <div className={styles.cartItemPrice}>${item.price}</div>
                    <div className={styles.quantityControls}>
                      <button
                        className={styles.quantityBtn}
                        onClick={() => handleDecreaseQuantity(item.id)}
                      >
                        -
                      </button>
                      <span className={styles.quantity}>{item.quantity}</span>
                      <button
                        className={styles.quantityBtn}
                        onClick={() => handleIncreaseQuantity(item.id)}
                      >
                        +
                      </button>
                      <button
                        className={styles.quantityBtn}
                        onClick={() => handleRemoveItem(item.id)}
                        style={{
                          marginLeft: "10px",
                          background: "#dc3545",
                          color: "white",
                        }}
                      >
                        ðŸ—‘
                      </button>
                    </div>
                  </div>
                </div>
              ))}
            </div>

            <div className={styles.cartTotal}>
              <div className={styles.totalText}>
                Total: ${totalAmount.toFixed(2)}
              </div>
              <button className={styles.checkoutBtn} onClick={handleCheckout}>
                Checkout
              </button>
            </div>
          </>
        )}
      </div>
    </>
  );
});

Cart.displayName = "Cart";

export default Cart;
