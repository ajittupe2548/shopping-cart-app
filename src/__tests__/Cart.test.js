import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import Cart from "../components/Cart";
import cartReducer from "../store/cartSlice";

const renderWithRedux = (component, initialState = {}) => {
  const store = configureStore({
    reducer: {
      cart: cartReducer,
    },
    preloadedState: initialState,
  });

  return {
    ...render(<Provider store={store}>{component}</Provider>),
    store,
  };
};

// Mock window.alert
global.alert = jest.fn();

describe("Cart", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("shows empty cart message when cart is empty", () => {
    const initialState = {
      cart: {
        items: [],
        totalQuantity: 0,
        totalAmount: 0,
        isOpen: true,
      },
    };

    renderWithRedux(<Cart />, initialState);

    expect(screen.getByText("Your cart is empty")).toBeInTheDocument();
  });

  test("displays cart items correctly", () => {
    const initialState = {
      cart: {
        items: [
          {
            id: 1,
            title: "Test Product 1",
            price: 29.99,
            image: "test1.jpg",
            quantity: 2,
          },
          {
            id: 2,
            title: "Test Product 2",
            price: 39.99,
            image: "test2.jpg",
            quantity: 1,
          },
        ],
        totalQuantity: 3,
        totalAmount: 99.97,
        isOpen: true,
      },
    };

    renderWithRedux(<Cart />, initialState);

    expect(screen.getByText("Test Product 1")).toBeInTheDocument();
    expect(screen.getByText("Test Product 2")).toBeInTheDocument();
    expect(screen.getByText("$29.99")).toBeInTheDocument();
    expect(screen.getByText("$39.99")).toBeInTheDocument();
    expect(screen.getByText("Total: $99.97")).toBeInTheDocument();
  });

  test("increases item quantity when + button is clicked", () => {
    const initialState = {
      cart: {
        items: [
          {
            id: 1,
            title: "Test Product",
            price: 29.99,
            image: "test.jpg",
            quantity: 1,
          },
        ],
        totalQuantity: 1,
        totalAmount: 29.99,
        isOpen: true,
      },
    };

    const { store } = renderWithRedux(<Cart />, initialState);

    const increaseButton = screen.getByText("+");
    fireEvent.click(increaseButton);

    const state = store.getState();
    expect(state.cart.items[0].quantity).toBe(2);
  });

  test("decreases item quantity when - button is clicked", () => {
    const initialState = {
      cart: {
        items: [
          {
            id: 1,
            title: "Test Product",
            price: 29.99,
            image: "test.jpg",
            quantity: 2,
          },
        ],
        totalQuantity: 2,
        totalAmount: 59.98,
        isOpen: true,
      },
    };

    const { store } = renderWithRedux(<Cart />, initialState);

    const decreaseButton = screen.getByText("-");
    fireEvent.click(decreaseButton);

    const state = store.getState();
    expect(state.cart.items[0].quantity).toBe(1);
  });

  test("removes item when remove button is clicked", () => {
    const initialState = {
      cart: {
        items: [
          {
            id: 1,
            title: "Test Product",
            price: 29.99,
            image: "test.jpg",
            quantity: 1,
          },
        ],
        totalQuantity: 1,
        totalAmount: 29.99,
        isOpen: true,
      },
    };

    const { store } = renderWithRedux(<Cart />, initialState);

    const removeButton = screen.getByText("ðŸ—‘");
    fireEvent.click(removeButton);

    const state = store.getState();
    expect(state.cart.items).toHaveLength(0);
  });

  test("handles checkout correctly", () => {
    const initialState = {
      cart: {
        items: [
          {
            id: 1,
            title: "Test Product",
            price: 29.99,
            image: "test.jpg",
            quantity: 1,
          },
        ],
        totalQuantity: 1,
        totalAmount: 29.99,
        isOpen: true,
      },
    };

    const { store } = renderWithRedux(<Cart />, initialState);

    const checkoutButton = screen.getByText("Checkout");
    fireEvent.click(checkoutButton);

    expect(global.alert).toHaveBeenCalledWith(
      "Thank you for your purchase! Total: $29.99"
    );

    const state = store.getState();
    expect(state.cart.items).toHaveLength(0);
    expect(state.cart.isOpen).toBe(false);
  });

  test("closes cart when close button is clicked", () => {
    const initialState = {
      cart: {
        items: [],
        totalQuantity: 0,
        totalAmount: 0,
        isOpen: true,
      },
    };

    const { store } = renderWithRedux(<Cart />, initialState);

    const closeButton = screen.getByText("âœ•");
    fireEvent.click(closeButton);

    const state = store.getState();
    expect(state.cart.isOpen).toBe(false);
  });

  test("matches snapshot when cart is closed", () => {
    const initialState = {
      cart: {
        items: [],
        totalQuantity: 0,
        totalAmount: 0,
        isOpen: false,
      },
    };

    const { container } = renderWithRedux(<Cart />, initialState);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("matches snapshot when cart is open with items", () => {
    const initialState = {
      cart: {
        items: [
          {
            id: 1,
            title: "Test Product",
            price: 29.99,
            image: "test.jpg",
            quantity: 2,
          },
        ],
        totalQuantity: 2,
        totalAmount: 59.98,
        isOpen: true,
      },
    };

    const { container } = renderWithRedux(<Cart />, initialState);
    expect(container.firstChild).toMatchSnapshot();
  });
});
