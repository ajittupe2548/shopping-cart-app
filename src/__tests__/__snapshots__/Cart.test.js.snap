// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Cart matches snapshot when cart is closed 1`] = `
<div
  class="cartOverlay "
/>
`;

exports[`Cart matches snapshot when cart is open with items 1`] = `
<div
  class="cartOverlay open"
/>
`;
